apps/ex_wire/lib/ex_wire/config.ex

```diff
@@ -1,96 +1,123 @@
         key
 
       :random ->
         ECDH.new_ecdh_keypair()
         |> Tuple.to_list()
         |> List.last()
+        |> set_env!(:private_key)
     end
   end
 
-  @spec udp_network_adapter() :: {atom(), atom()}
-  def udp_network_adapter do
-    node_discovery_params()[:network_adapter]
+  @spec udp_network_adapter(Keyword.t()) :: {atom(), atom()}
+  def udp_network_adapter(given_params \\ []) do
+    node_discovery_params(given_params, :network_adapter)
   end
 
   @spec public_key() :: binary()
-  def public_key do
+  def public_key() do
     {:ok, public_key} = Signature.get_public_key(private_key())
 
     public_key
   end
 
-  @spec discovery() :: boolean()
-  def discovery do
-    get_env(:discovery)
+  @spec public_node_url() :: String.t()
+  def public_node_url() do
+    my_public_key_encoded =
+      public_key()
+      |> Base.encode16(case: :lower)
+
+    my_port = listen_port()
+    # TODO: This is only valid for IPv4 addresses
+    my_public_ip =
+      public_ip()
+      |> Enum.map(fn digit -> to_string(digit) end)
+      |> Enum.join(".")
+
+    "enode://#{my_public_key_encoded}@#{my_public_ip}:#{my_port}"
+  end
+
+  @spec perform_discovery?(Keyword.t()) :: boolean()
+  def perform_discovery?(given_params \\ []) do
+    get_env(given_params, :discovery, false)
   end
 
-  @spec public_ip() :: [integer()]
-  def public_ip do
-    get_env(:public_ip) || [127, 0, 0, 1]
+  @spec public_ip(Keyword.t()) :: [integer()]
+  def public_ip(given_params \\ []) do
+    get_env(given_params, :public_ip, @default_public_ip)
   end
 
   @spec node_id() :: ExWire.node_id()
-  def node_id do
+  def node_id() do
     Crypto.node_id_from_public_key(public_key())
   end
 
-  @spec listen_port() :: integer()
-  def listen_port do
-    get_env(:port) || 30_304
+  @spec listen_port(Keyword.t()) :: integer()
+  def listen_port(given_params \\ []) do
+    node_discovery_params(given_params, :port, @default_port)
   end
 
-  @spec protocol_version() :: integer()
-  def protocol_version do
-    get_env(:protocol_version)
+  @spec protocol_version(Keyword.t()) :: integer()
+  def protocol_version(given_params \\ []) do
+    get_env(given_params, :protocol_version)
   end
 
-  @spec network_id() :: integer()
-  def network_id do
-    get_env(:network_id)
+  @spec network_id(Keyword.t()) :: integer()
+  def network_id(given_params \\ []) do
+    get_env(given_params, :network_id)
   end
 
-  @spec p2p_version() :: integer()
-  def p2p_version do
-    get_env(:p2p_version)
+  @spec p2p_version(Keyword.t()) :: integer()
+  def p2p_version(given_params \\ []) do
+    get_env(given_params, :p2p_version)
   end
 
-  @spec caps() :: [{String.t(), integer()}]
-  def caps do
-    get_env(:caps)
+  @spec caps(Keyword.t()) :: [{String.t(), integer()}]
+  def caps(given_params \\ []) do
+    get_env(given_params, :caps)
   end
 
   @spec client_id() :: String.t()
-  def client_id do
+  def client_id() do
     version = Mix.Project.config()[:version]
```

Comment:
this won't work in a release... should this be part of your PR or addressed later?

Reply:
Hm, yeah, interesting question. We haven't discussed this yet! A release configuration setts it's version:
`
release :mana do
  set version: "0.1.0"
  set applications: [
    :runtime_tools,
    blockchain: :permanent,
    cli: :permanent,
    evm: :permanent,
    ex_wire: :permanent,
    exth_crypto: :permanent,
    merkle_patricia_tree: :permanent
  ]
end
`
Now this part of configuration get's evaluated when building so you could potentially set it to:
set version: "#{System.get_env("RELEASE_VERSION")}" or even pull it out of a file or... whatever. Maybe, for consistency reasons having a file in root/priv isn't a bad solution? So if a git tag is a release, that commit needs to update the file too. WDYT?